examples/pytorch/mnist-ddp/mnist.ipynb

```diff
@@ -0,0 +1,627 @@
+    "!pip install torch==2.5.1\n",
+    "!pip install torchvision==0.20.0"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Define the training function"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 40,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def train_fashion_mnist(params):\n",
+    "    import os\n",
+    "\n",
+    "    import torch\n",
+    "    import torch.distributed as dist\n",
+    "    import torch.nn.functional as F\n",
+    "    import torch.nn as nn\n",
+    "    import torchvision.transforms as transforms\n",
+    "    from torch.nn.parallel import DistributedDataParallel\n",
+    "    from torch.optim.lr_scheduler import StepLR\n",
+    "    from torch.utils.data import DataLoader\n",
+    "    from torch.utils.data.distributed import DistributedSampler\n",
+    "    from torchvision.datasets import FashionMNIST\n",
+    "\n",
+    "    # Define the PyTorch CNN model to be trained\n",
+    "    class Net(nn.Module):\n",
+    "        def __init__(self):\n",
+    "            super(Net, self).__init__()\n",
+    "            self.conv1 = nn.Conv2d(1, 20, 5, 1)\n",
+    "            self.conv2 = nn.Conv2d(20, 50, 5, 1)\n",
+    "            self.fc1 = nn.Linear(4 * 4 * 50, 500)\n",
+    "            self.fc2 = nn.Linear(500, 10)\n",
+    "\n",
+    "        def forward(self, x):\n",
+    "            x = F.relu(self.conv1(x))\n",
+    "            x = F.max_pool2d(x, 2, 2)\n",
+    "            x = F.relu(self.conv2(x))\n",
+    "            x = F.max_pool2d(x, 2, 2)\n",
+    "            x = x.view(-1, 4 * 4 * 50)\n",
+    "            x = F.relu(self.fc1(x))\n",
+    "            x = self.fc2(x)\n",
+    "            return F.log_softmax(x, dim=1)\n",
+    "\n",
+    "    # Use NCCL is a GPU is available, otherwise use Gloo as communication backend\n",
+    "    device, backend = (\"cuda\", \"nccl\") if torch.cuda.is_available() else (\"cpu\", \"gloo\")\n",
+    "\n",
+    "    print(f\"Using Device: {device}, Backend: {backend}\")\n",
+    "\n",
+    "    # Setup PyTorch Distributed\n",
+    "    local_rank = int(os.getenv(\"LOCAL_RANK\", 0))\n",
+    "    dist.init_process_group(backend=backend)\n",
+    "\n",
+    "    print(\n",
+    "        \"Distributed Training for WORLD_SIZE: {}, RANK: {}, LOCAL_RANK: {}\".format(\n",
+    "            dist.get_world_size(),\n",
+    "            dist.get_rank(),\n",
+    "            local_rank,\n",
+    "        )\n",
+    "    )\n",
+    "\n",
+    "    # Create the model and load it into the device\n",
+    "    device = torch.device(f\"{device}:{local_rank}\")\n",
+    "    model = DistributedDataParallel(Net().to(device))\n",
+    "\n",
+    "    # Retrieve the Fashion-MNIST dataset\n",
+    "    dataset = FashionMNIST(\n",
+    "        \"./data\",\n",
+    "        train=True,\n",
+    "        download=True,\n",
+    "        transform=transforms.Compose([transforms.ToTensor()]),\n",
```

Comment:
One thing that I notice is that we download data incorrectly in our V1 example: https://github.com/kubeflow/training-operator/blob/df640a46fa541ffdddb70e0a6a7744280c93686f/examples/pytorch/image-classification/Train-CNN-with-FashionMNIST.ipynb.

If we run this code in Multi-Proc Multi-Node environment, we should only download data when:
`LOCAL_RANK == 0` 
After that, we should use `dist.barrier()` to wait until dataset will be available in every PyTorch Node, so we can distribute it with `DistributedSampler`.
Otherwise, it downloads data multiple times inside the same PyTorch node.

@astefanutti Would be nice if we could fix it in V2.

cc @kubeflow/wg-training-leads @Electronic-Waste @shravan-achar @kuizhiqing @seanlaii @Syulin7

Reply:
I'd be inclined to stick to single process node to keep that example as simple as possible.

That seems a good scenario for a more advanced example that'd cover multi-processes node.

Reply:
Sure, maybe we can start with single-proc per node in this example.
@astefanutti Should we also use only CPUs in this example so users can quickly replicate it locally without GPU?
Although, The problem with CPU is that PyTorch sets nproc_per_node to the actual number of available CPUs if we don't override it: https://github.com/pytorch/pytorch/blob/main/torch/distributed/run.py#L650

Reply:
I agree, we can comment out the GPU extended resource.

For PyTorch, that's a good point, is PyTorch actually cgroups aware? If yes then passing `"cpu": 1` to the Train API should be enough.

Looking at the code, it means whether `os.cpu_count()` is cgroups aware or not, which may be the case according to https://github.com/python/cpython/issues/80235, but possibly only in recent Python versions.

So it seems setting `nproc_per_node` as you've suggested is the safest option.

Reply:
I don't think it is cgroup aware.
Should we set nproc_per_node = number of CPUs if `nproc_per_node` is not set by the user ?

Reply:
There is also an issue when no GPU is requested that makes all the GPUs visible from the workers by default unless `NVIDIA_VISIBLE_DEVICES=none` is set: https://github.com/NVIDIA/k8s-device-plugin/issues/61.

Not sure how often a user would want to run a job with CPU-only on nodes with GPUs, but if that's the case then that's an issue.

Reply:
Oh, I see. I just thought that if we set only CPUs there, the workloads will be assigned to the CPU-based Kubernetes Node, since we will not add the NVIDIA taint and tolerations.

Reply:
Right, for nodes with GPUs that haven't been tainted, CPU-only Pods will still be scheduled on these nodes. `NVIDIA_VISIBLE_DEVICES=none` can be added to the training runtime environment variables to mitigate that if needed.

Reply:
> Should we set nproc_per_node = number of CPUs if nproc_per_node is not set by the user ?

I think so. That becomes very problematic with large nodes that has lots of CPUs.

Now if CPU resources is not set in `resources_per_node`, that's still a problem when `numProcPerNode` in the training runtime ML policy is set to `auto` for PyTorch. Maybe we can override `nproc_per_node` and set it to one in that case.